sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -0,0 +1,410 @@
+
+mutex_t _nib_mutex = MUTEX_INIT;
+
+static void _override_node(const ipv6_addr_t *addr, unsigned iface,
+                           _nib_t *node);
+static inline bool _node_unreachable(_nib_t *node);
+
+void _nib_init(void)
+{
+    _prime_def_router = NULL;
+    _last_def_router = NULL;
+    _next_removable.next = NULL;
+    memset(_nodes, 0, sizeof(_nodes));
+    memset(_def_routers, 0, sizeof(_def_routers));
+    memset(_nis, 0, sizeof(_nis));
+    /* TODO: load ABR information from persistent memory */
+}
+
+_nib_t *_nib_alloc(const ipv6_addr_t *addr, unsigned iface)
+{
+    _nib_t *node = NULL;
+
+    assert(addr != NULL);
+    DEBUG("nib: Allocating on-link node entry (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+    for (unsigned i = 0; i < GNRC_IPV6_NIB_NUMOF; i++) {
+        _nib_t *tmp = &_nodes[i];
+
+        if ((ipv6_addr_equal(addr, &tmp->ipv6)) &&
+            (_nib_get_if(tmp) == iface)) {
+            /* exact match */
+            DEBUG("  %p is an exact match\n", (void *)tmp);
+            return tmp;
+        }
+        if ((node == NULL) && (tmp->mode == _EMPTY)) {
+            node = tmp;
+        }
+    }
+    if (node != NULL) {
+        DEBUG("  using %p\n", (void *)node);
+        _override_node(addr, iface, node);
+    }
+#if ENABLE_DEBUG
+    else {
+        DEBUG("  NIB full\n");
+    }
+#endif
+    return node;
+}
+
+static inline bool _is_gc(_nib_t *nib)
+{
+    return ((nib->mode & ~(_NC)) == 0) &&
+            ((nib->info & GNRC_IPV6_NIB_NC_INFO_AR_STATE_MASK) ==
+              GNRC_IPV6_NIB_NC_INFO_AR_STATE_GC);
+}
+
+_nib_t *_nib_nc_add(const ipv6_addr_t *addr, unsigned iface, uint16_t cstate)
+{
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_DELAY);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_PROBE);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_REACHABLE);
+    _nib_t *nib = _nib_alloc(addr, iface);
+    if (nib != NULL) {
+        DEBUG("nib: Adding to neighbor cache (addr = %s, iface = %u)\n",
+              ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+        if (!(nib->mode & _NC)) {
+            nib->info &= ~GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK;
+            nib->info |= (cstate & GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK);
+            nib->mode |= _NC;
+        }
+        if (nib->next == NULL) {
+            DEBUG("nib: queueing (addr = %s, iface = %u) for potential removal\n",
+                  ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+            /* add to next removable list, if not already in it */
+            clist_rpush(&_next_removable, (clist_node_t *)nib);
+        }
+    }
+    else {
+        _nib_t *first = (_nib_t *)clist_lpop(&_next_removable);
+        _nib_t *tmp = first;
+
+        DEBUG("nib: Searching for replaceable entries (addr = %s, iface = %u)\n",
+              ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+        if (tmp != NULL) {
+            do {
```

Comment:
do we actually need the clist? Why don't you just iterate over the whole nib array until a garbage collectible entry is found? This would reduce the code complexity a lot and you would even lose the `next` pointer in the `_nib_t` struct.